### PR TITLE
Support CAN for Linux boards

### PR DESCRIPTION
**Pull request by AlexeyBulatov**
Hello all,

This pull request adds CAN driver and its initialization for Linux boards. Driver based on code from UAVCAN for linux (https://github.com/UAVCAN/libuavcan/blob/master/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp).

Best regards,
Alexey Bulatov.

### CODE REVIEW

**Comment by EShamaev** [comment]
May I ask to rework the PR based on https://github.com/ArduPilot/ardupilot/pull/6167 as chances are it will get merged and will be incompatible with this PR.

**Comment by AlexeyBulatov** [comment]
Rebased on #6167

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ NOINLINE void Copter::send_heartbeat(mavlink_channel_t chan)
 #endif
 
     // we are armed if we are not initialising
-    if (motors->armed()) {
+    if (motors != nullptr && motors->armed()) {
```
Comment (EShamaev) [inline ArduCopter/GCS_Mavlink.cpp]: How is this related to CAN?

**Comment by EShamaev** [comment]
Please advice if it was flight tested?

**Comment by staroselskii** [comment]
> Please advice if it was flight tested?

It was only bench tested. We'll try to flight test it on the next week. We don't expect any serious difficulties, though. The barometer did work as well as GNSS did.

Are we considering current master close to a stable release?

**Comment by EShamaev** [comment]
Current master is stable, but https://github.com/ArduPilot/ardupilot/pull/6167 is very close to being merged. As it changes current master a lot, I think it is better to use it a a base.

**Comment by EShamaev** [comment]
Can be rebased to master and fully tested :)

**Comment by AlexeyBulatov** [comment]
Rebased and successfully tested.

File: ArduCopter/GCS_Mavlink.cpp (hunk 1 above)
Comment (AlexeyBulatov) [inline ArduCopter/GCS_Mavlink.cpp]: Because of adding delay for magnetometer and barometer discovery in AP_BoardConfig, I get SEGFAULT since object motors doesn't created.

**Comment by AlexeyBulatov** [comment]
Ping

**Comment by staroselskii** [comment]
Hello guys!
We've flown it multiple times and really would definitely like to get a review on this one. Thanks!

**Comment by OXINARF** [comment]
It is on my list, but my backlog is huge. Sorry for the delay!

File: libraries/AP_HAL_Linux/CAN.cpp [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+        }
+
+        // Removing the frame from the queue even if transmission failed
+        _tx_queue.pop();
+    }
+}
+
+void LinuxCAN::_pollRead()
+{
+    while (true)
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/CAN.cpp]: can we get a counter here to only loop to a maximum number of iterations? This avoids situations in which we would be stuck processing packets here.

File: libraries/AP_HAL_Linux/CAN.cpp [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+        getsockopt(pfd.fd, SOL_SOCKET, SO_ERROR, reinterpret_cast<void*>(&error), &errlen);
+
+        down_ = error == ENETDOWN || error == ENODEV;
+
+        hal.console->printf("Iface %d is dead; error %d", this->getFileDescriptor(), error);
+    }
+}
+
+void LinuxCANDriver::register_thread() {
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&LinuxCANDriver::_timer_tick, void));
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/CAN.cpp]: I think we should have a real thread instead of hooking to the timer (see the Thread abstraction in the Linux HAL... PollerThread would be ideal for this since it has a full featured epoll abstraction).

The architecture here is a bit odd in which you register a callback here, then you call out to `p_uavcan->do_cyclic()`, which calls back into here... but it's doable with PollerThread, too. Long term I'd like to see users of timer to go away.

File: libraries/AP_HAL_Linux/CAN.h [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+namespace Linux {
+
+enum class SocketCanError
+{
+    SocketReadFailure,
+    SocketWriteFailure,
+    TxTimeout
+};
+
+class LinuxCAN: public AP_HAL::CAN {
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/CAN.h]: This is already inside the Linux namespace. We moved away from "Linux::Linux*" naming some time ago. I think this should be named simply CAN.

File: libraries/AP_HAL_Linux/CAN.h [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+    void initialized(bool val) override;
+
+    AP_UAVCAN *p_uavcan;
+
+    AP_UAVCAN *get_UAVCAN(void) override;
+
+    void set_UAVCAN(AP_UAVCAN *uavcan) override;
+
+    virtual LinuxCAN* getIface(uint8_t iface_index) override;
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/CAN.h]: is this override correct? I can't see a method in AP_HAL returning something of LinuxCAN type.

**Comment by magicrub** [comment]
What's the status on this? I need CAN support on Linux and would love this to go in (once it's ready, of course)

**Comment by staroselskii** [comment]
@magicrub 

We've been using this code for almost half a year now without any issues.

But of course we're planning to conform to all the comments. 

The work on the problems that have been pointed at above is planned for this week. So I guess, we'll make it by the end of the year.

**Comment by magicrub** [comment]
@staroselskii great, thanks! I'm very eager to use this so I'm motivated to help where possible. Can I do some of the updates for you and you fetch from me? I have another branch I'll be putting on top of this so I'll adapting that to this PR too

**Comment by staroselskii** [comment]
@magicrub 

Sure thing! That is the most awesome scenario I can think of.

File: libraries/AP_HAL_Linux/CAN.h (hunk 5 above)
Comment (Programmer74) [reply libraries/AP_HAL_Linux/CAN.h]: Well, this is slightly incorrect. This has to return an ICanIface* type. The method is not defined in AP_HAL, this is defined in uavcan module [here](https://github.com/ArduPilot/uavcan/blob/9d75eba870e3b3935346229267ecee2c499f6575/libuavcan/include/uavcan/driver/can.hpp#L211).

**Comment by magicrub** [comment]
I've been chatting with @Programmer74 and he said he'll be making a new PR with all the changes.
